### PR TITLE
Allow editing networks via UI

### LINKED ISSUE #179: Renaming networks
It would have been good to rename an already saved network over the web client.
Especially for multi-user environments not everyone has access to the file system.

Sometimes you might want to change the network name or user name.
I think instead of deleting it, adding it again, trying to take a not of your channels etc., just like when we are adding a new network, having an option to change all these would be great.

### LINKED ISSUE #1161: Edit user commands from interface
- Commands should be editable from connect menu
- Make possible to open connect menu for existing networks to change settings

### PR DESCRIPTION
**Pull request by xPaw**
Some of the network code moved, and the editing code had to be done in such a way to apply same restrictions as adding a new connection. So this needs a lot of testing.

- [x] Fixes #179
- [x] Fixes #1161 (and practically #584, but the UI only exists when editing, not connecting, which is fine by me)
- [x] Update lobby name in UI if changed
- [x] Send `/nick` if nick changes
- [x] Update IRC object with new configuration
- [x] Look if `irc.user` variables need update (e.g. nick while disconnected)
- [x] Default nick to config default instead of hardcoded string

### CODE REVIEW

**Comment by astorije** [comment]
Wow, this is dope!

I haven't looked at the code yet, but to quickly discuss the overall behavior, this is why I would expect:

- If I change host/port/etc., it reconnects
- If I change the network name, it changes the name in the channel list UI (but doesn't reconnect if there is no change in host/port/etc.)
- If I change the nick, it sends a `/nick ...` (unless a full reconnect is necessary). This last bit is cool because we can then completely remove the nick editor UI from the message input!

UI can be made in a way that this behavior specificity is totally free of surprises (change the button to "Reconnect" instead of "Save" when reconnection will appear, extra info message above the save button, etc. several patterns to play with, and I'm happy to help wherever I can)

What are your thoughts on this?

**Comment by xPaw** [comment]
@astorije All the points you point out are indeed achievable. However I want to test out current implementation for unexpected surprises and bugs (because a lot of code was moved around) before adding more code.

**Comment by astorije** [comment]
That's fair. What is the expected behavior when hitting the submit button? i.e. what is it supposed to do and not to do?

**Comment by xPaw** [comment]
@astorije It should update in-memory and on-disk values. Keep in mind that this PR also moves the code around for just connect, so both should work as expected.

I'll have another commit soon that implements `/nick` if it's changed, so only changing lobby name and reconnecting will be left to do.

**Comment by MaxLeiter** [comment]
On mobile so can’t edit, #584 *

**Comment by xPaw** [comment]
@MaxLeiter I think there's hardly any use case where you'd know which commands to enter when creating a network. And it's hidding a bit of clutter as to not confuse users (e.g. on public instances).

**Comment by xPaw** [comment]
I've got all the UI stuff fleshed out. The only point that I didn't implement is auto reconnecting.

File: client/js/lounge.js [hunk 1]
```diff
@@ -141,6 +141,12 @@ $(function() {
 					text: "Join a channel…",
 					data: target.data("id"),
 				});
+				output += templates.contextmenu_item({
+					class: "edit",
+					action: "edit",
+					text: "Edit network",
+					data: target.closest(".network").data("uuid"),
+				});
```
Comment (astorije) [inline client/js/lounge.js]: Mind putting this right after the network name, and calling it "Edit this network..."?

**Comment by astorije** [comment]
Do you think you could add a few tests to this? That would be pretty awesome.

**Comment by xPaw** [comment]
Added some tests.

**Comment by McInkay** [comment]
This will need to be refactored now that #1878 is merged, and use the new structure for adding context menu items.

**Comment by astorije** [comment]
@xPaw, what do you mean by:

> Default nick to config default instead of hardcoded string

**Comment by xPaw** [comment]
@astorije default nick fallback, what we discussed in #2297

File: src/helper.js [hunk 2]
```diff
@@ -176,6 +177,14 @@ function passwordCompare(password, expected) {
 	return bcrypt.compare(password, expected);
 }
 
+function getDefaultNick() {
+	if (!this.config.defaults.nick) {
+		return "thelounge";
+	}
```
Comment (astorije) [inline src/helper.js]: Is this default not supposed to come from our own default copy of the config file?

File: client/js/contextMenuFactory.js [hunk 3]
```diff
@@ -153,6 +153,22 @@ function addFocusItem() {
 	});
 }
 
+function addEditNetworkItem() {
+	function edit(itemData) {
+		socket.emit("network:get", itemData);
+		$("#connect").text("Loading");
```
Comment (astorije) [inline client/js/contextMenuFactory.js]: Is this loading text legacy from a previous attempt or something? I couldn't manage to trigger it, even when emulating slow network and even when simulating lost connection.

However, I do notice a flash when clicking on "Edit this network", for a very short window (100ms?) the fields are filled with the connect defaults, then change to the actual values. I'm wondering if we can keep them empty until they're filled.

File: src/helper.js (hunk 2 above)
Comment (xPaw) [reply src/helper.js]: This is a fail-safe if user overrides the default with empty string.

**Comment by astorije** [comment]
@xPaw, sorry about that but I just noticed while playing a bit more with your branch: going to the edit form appends the `#connect` hash to the URL, so reloading the page and browser history don't work, mind looking into that?

Also, I see you amended your commits, did you manage to do something about the "flash" when going to edit form? I just tried and I still see it, can't see the updated `Loading...`.

Also, not asking if that's on purpose, but that PR has evolved since we first talked about it so double-checking: if I change the host/port, I don't get re-connected, this is intended behavior for this PR, correct?

Feel free to assign me to the PR when you're done with changes!

**Comment by xPaw** [comment]
> sorry about that but I just noticed while playing a bit more with your branch: going to the edit form appends the #connect hash to the URL, so reloading the page and browser history don't work, mind looking into that?

> Also, I see you amended your commits, did you manage to do something about the "flash" when going to edit form?

Both the flash and `#connect` are related issues because I'm re-using the connect window. It's a bit painful to create a separate window for editing networks that would work with page reloads and wouldn't involve a bunch of nasty code. So I just chose to keep it this way for now.

> Also, not asking if that's on purpose, but that PR has evolved since we first talked about it so double-checking: if I change the host/port, I don't get re-connected, this is intended behavior for this PR, correct?

Yep intended.

**Comment by astorije** [comment]
> It's a bit painful to create a separate window for editing networks that would work with page reloads and wouldn't involve a bunch of nasty code. So I just chose to keep it this way for now.

I understand and that's fine for now I think 👍 (though maybe create an issue and tag it to v3 to keep track of this?)

Is there any chance you could fix the URL though, at least? It doesn't have to be the most perfect code if in the long-run we're having a separate window or something. That would fix the back/forward buttons, I believe.

**Comment by MaxLeiter** [comment]
xPaw: could you assign a different hash when the window is loaded, and add s temporary check for it on load? 

On second thought, don’t really think it’s worth the trouble

**Comment by xPaw** [comment]
> Is there any chance you could fix the URL though, at least?

That directly relates to window management sadly. Trying to rewrite the url will effectively mean I will have to manage the page reloading/popstate to keep track of it, which @MaxLeiter suggested.

**Pull request merged**